traffic_ops/app/lib/API/Server.pm

```diff
@@ -569,10 +575,17 @@ sub get_unassigned_servers_by_dsid {
 	$ds_server_criteria{'deliveryservice.id'} = $ds_id;
 
 	my @assigned_servers;
-	if ( &is_privileged($self) || $self->is_delivery_service_assigned($ds_id) ) {
+	my $tenant_utils = Utils::Tenant->new($self);
+	my $tenants_data = $tenant_utils->create_tenants_data_from_db();
+
+	if (!$tenant_utils->is_ds_resource_accessible($tenants_data, $ds->tenant_id)) {
```

Comment:
Seems ok to me. 

case 1: "use=tenancy" == true, and resource is accessible to tenant:  
1. "if" test is false, continue
2. "elseif" test is true (as ignore_ds_users_table==true), so we get the list of servers

case 2: "use=tenancy" == true, and resource is not accessible to tenant:  
1. "if" test is true, return forbidden

case 3: "use=tenancy" == false, and ds is assigned to user
1. "if" test is false, continue
2. "elseif" test is true (as the ds is assigned to user), so we get the list of servers


case 3: "use=tenancy" == false, and ds is not assigned to user
1. "if" test is false, continue
2. "elseif" test is false (as the ds is not assigned to user),
3. We reach the failure